docs/src/main/asciidoc/websockets-next-reference.adoc

```diff
@@ -828,60 +893,108 @@ public class Endpoint {
 <1> The echo callback method can only be invoked if the current security identity has an `admin` role.
 <2> The error handler is invoked in case of the authorization failure.
 
-`SecurityIdentity` is initially created during a secure HTTP upgrade and associated with the websocket connection.
+==== Secure server endpoints with permission checkers
 
-NOTE: When OpenID Connect extension is used and token expires, Quarkus automatically closes connection.
+WebSocket endpoints can be secured with the xref:security-authorize-web-endpoints-reference.adoc#permission-checker[permission checkers].
+We recommend to <<secure-http-upgrade>> rather than individual endpoint methods. For example:
```

Comment:
I'm not sure about the recommendation. What if one requires an authentication at the upgrade time, and different permissions for different endpoint methods ?

Reply:
For now, there is exactly one endpoint for one connection.

Reply:
As for endpoint methods, well, if you can open connection I think you can also get messages etc.

Reply:
Sorry, looks like I'm missing something. Does it work now:

```
Authenticated <1>
@WebSocket(path = "/end")
public class Endpoint {
    @Inject
    SecurityIdentity currentIdentity;
    @OnOpen
    String open() {
        return "ready";
    }
    @RolesAllowed("admin")
    @OnTextMessage
    String echo(String message) {
        return message;
    }
}
```
?

Reply:
It does work, but every time `echo` is invoked, allowed roles are checked. The `SecurityIdentity` roles are immutable, therefore I'd check them once instead of many times.

Reply:
I'll put it another way @sberyozkin , because maybe it is me who is missing something. What do you get by annotating `echo` with `@RolesAllowed` instead of `Endpoint`. do you have in mind a scenario that would profit from that?